OpenSim/Common/APDMDataReader.cpp

```diff
@@ -0,0 +1,243 @@
+    // Tables could be empty if data is not present in file(s)
+    DataAdapter::OutputTables tables{};
+    auto orientationTable = std::make_shared<TimeSeriesTableQuaternion>(times, rotationsData, labels);
+    orientationTable->updTableMetaData()
+        .setValueForKey("DataRate", std::to_string(dataRate));
+    tables.emplace(IMUDataUtilities::Orientations, orientationTable);
+
+    std::vector<double> emptyTimes;
+    auto accelerationTable = (foundLinearAccelerationData ?
+        std::make_shared<TimeSeriesTableVec3>(times, linearAccelerationData, labels) :
+        std::make_shared<TimeSeriesTableVec3>(emptyTimes, linearAccelerationData, labels));
+    accelerationTable->updTableMetaData()
+        .setValueForKey("DataRate", std::to_string(dataRate));
+    tables.emplace(IMUDataUtilities::LinearAccelerations, accelerationTable);
+
+    auto magneticHeadingTable = (foundMagneticHeadingData ?
+        std::make_shared<TimeSeriesTableVec3>(times, magneticHeadingData, labels) :
+        std::make_shared<TimeSeriesTableVec3>(emptyTimes, magneticHeadingData, labels));
+    magneticHeadingTable->updTableMetaData()
+        .setValueForKey("DataRate", std::to_string(dataRate));
+    tables.emplace(IMUDataUtilities::MagneticHeading, magneticHeadingTable);
+
+    auto angularVelocityTable = (foundAngularVelocityData ?
+        std::make_shared<TimeSeriesTableVec3>(times, angularVelocityData, labels) :
+        std::make_shared<TimeSeriesTableVec3>(emptyTimes, angularVelocityData, labels));
+    angularVelocityTable->updTableMetaData()
+        .setValueForKey("DataRate", std::to_string(dataRate));
+    tables.emplace(IMUDataUtilities::AngularVelocity, angularVelocityTable);
+
+    return tables;
+}
+
+void APDMDataReader::find_start_column(std::vector<std::string> tokens, 
+    std::vector<std::string> search_labels,
+    const std::string& sensorName,
+    std::vector<int>& indices) const {
+    // Search for "sensorName/{search_labels} in tokens, append result to indices if found"
+    std::string firstLabel = sensorName + search_labels[0];
+    // look for first label, when found check/confirm the rest. Out of order is not supported
+    int found_index = -1;
+    std::vector<std::string>::iterator it = std::find(tokens.begin(), tokens.end(), firstLabel);
+    if (it != tokens.end()) {
+        found_index = static_cast<int>(std::distance(tokens.begin(), it));
+        // now check the following indices for match with remaining search_labels 
+        bool match = true;
+        for (int remaining = 1; remaining < search_labels.size() && match; remaining++) {
+            match = tokens[found_index + remaining].
+                compare(sensorName + search_labels[remaining]) == 0;
+        }
+        if (match) {
+            indices.push_back(found_index);
+                return;
+            }
+            else { // first label found but the remaining didn't. Warn
```

Comment:
"first label found but the remaining labels not found". Where is the warning?

Reply:
Good find @chrisdembia , I was going to write warning to stdout but other readers/adapters I saw tend to throw which is more aggressive. What would you suggest?

Reply:
If there is an error in the file, there should be an exception.